sw/device/silicon_creator/lib/epmp.h

```diff
@@ -0,0 +1,216 @@
+   *
+   * Note: these are the low 32 bits of mseccfg only. The high 32 bits
+   * are set to 0.
+   */
+  uint32_t mseccfg;
+} epmp_state_t;
+
+/**
+ * Configure the given PMP entry in state using the Top-Of-Range addressing
+ * mode.
+ *
+ * @param state The ePMP state to update.
+ * @param entry The index of the entry to update.
+ * @param region The memory region to encode in the address registers.
+ * @param perm The permissions to set for the entry.
+ */
+inline void epmp_state_configure_tor(epmp_state_t *state, uint32_t entry,
+                                     epmp_region_t region, epmp_perm_t perm) {
+  // Set address registers.
+  if (entry > 0) {
+    state->pmpaddr[entry - 1] = region.start >> 2;
+  }
+  state->pmpaddr[entry] = region.end >> 2;
+
+  // Set configuration register.
+  bitfield_field32_t field = {.mask = 0xff, .index = (entry % 4) * 8};
+  state->pmpcfg[entry / 4] = bitfield_field32_write(state->pmpcfg[entry / 4],
+                                                    field, kEpmpModeTor | perm);
+}
+
+/**
+ * Configure the given PMP entry in state using the Naturally-Aligned-4-byte
+ * addressing mode.
+ *
+ * @param state The ePMP state to update.
+ * @param entry The index of the entry to update.
+ * @param region The memory region to encode in the address registers.
+ * @param perm The permissions to set for the entry.
+ */
+inline void epmp_state_configure_na4(epmp_state_t *state, uint32_t entry,
+                                     epmp_region_t region, epmp_perm_t perm) {
+  // Set address register.
+  state->pmpaddr[entry] = region.start >> 2;
+
+  // Set configuration register.
+  bitfield_field32_t field = {.mask = 0xff, .index = (entry % 4) * 8};
+  state->pmpcfg[entry / 4] = bitfield_field32_write(state->pmpcfg[entry / 4],
+                                                    field, kEpmpModeNa4 | perm);
+}
+
+/**
+ * Configure the given PMP entry in state using the
+ * Naturally-Aligned-Power-Of-Two addressing mode.
+ *
+ * @param state The ePMP state to update.
+ * @param entry The index of the entry to update.
+ * @param region The memory region to encode in the address registers.
+ * @param perm The permissions to set for the entry.
+ */
+inline void epmp_state_configure_napot(epmp_state_t *state, uint32_t entry,
+                                       epmp_region_t region, epmp_perm_t perm) {
+  // Set address register.
+  uint32_t len = (region.end - region.start - 1) >> 3;
+  state->pmpaddr[entry] = (region.start >> 2) | len;
+
+  // Set configuration register.
+  bitfield_field32_t field = {.mask = 0xff, .index = (entry % 4) * 8};
+  state->pmpcfg[entry / 4] = bitfield_field32_write(
+      state->pmpcfg[entry / 4], field, kEpmpModeNapot | perm);
+}
+
+/**
+ * Report whether the given state matches the current hardware ePMP
+ * configuration.
+ *
+ * @param state Expected values of the ePMP CSRs.
+ */
+bool epmp_state_check(const epmp_state_t *state);
```

Comment:
Done. Can you check the new error code @alphan?

Reply:
In particular do you think 'MemoryProtect' is a good prefix or should we use 'Epmp' or something else?

Reply:
I think Epmp for consistency. Maybe `kErrorEpmpBadState`?

Reply:
Changed to `Epmp`. I left it as `BadCheck` for now. I did consider `BadState` originally but the failure doesn't necessarily indicate that the `state` is wrong (it could just as easily be the hardware configuration). `kErrorEpmpMismatch` maybe?

Reply:
Good point, `BadCheck` SGTM.